pkg/ingester/ingester_v2_backfill.go

```diff
@@ -0,0 +1,711 @@
+	buckets := i.TSDBState.backfillDBs.getBucketsForUser(userID)
+	if buckets == nil {
+		return nil, nil
+	}
+	var queriers []storage.Querier
+	defer func() {
+		for _, q := range queriers {
+			q.Close()
+		}
+	}()
+	buckets.RLock()
+	for _, b := range buckets.buckets {
+		if !b.overlaps(from, through) {
+			mint, err := b.db.DB.StartTime()
+			if err != nil {
+				buckets.RUnlock()
+				return nil, err
+			}
+			maxt := b.db.Head().MaxTime()
+			if !overlapsOpenInterval(mint, maxt, from, through) {
+				continue
+			}
+		}
+
+		q, err := b.db.Querier(ctx, from, through)
+		if err != nil {
+			buckets.RUnlock()
+			return nil, err
+		}
+
+		queriers = append(queriers, q)
+	}
+	buckets.RUnlock()
+
+	if len(queriers) == 0 {
+		return nil, nil
+	}
+
+	result := make([]storage.SeriesSet, len(queriers))
+	errC := make(chan error, 1)
+	var wg sync.WaitGroup
+	for i, q := range queriers {
+		wg.Add(1)
+		go func(i int, q storage.Querier) {
+			defer wg.Done()
+
+			ss := q.Select(false, nil, matchers...)
+			if ss.Err() != nil {
+				select {
+				case errC <- ss.Err():
+				default:
+				}
+			}
+			result[i] = ss
+		}(i, q)
+	}
+
+	wg.Wait()
+	select {
+	case err := <-errC:
+		return nil, err
+	default:
+	}
+
+	return result, nil
+}
+
+func (i *Ingester) closeAllBackfillTSDBs() {
+	// Snapshotting of in-memory chunks can be considered as a small compaction, hence
+	// using that concurrency.
+	i.runConcurrentBackfillWorkers(context.Background(), i.cfg.BlocksStorageConfig.TSDB.HeadCompactionConcurrency, func(db *userTSDB) {
+		if err := db.Close(); err != nil {
+			level.Warn(util.Logger).Log("msg", "unable to close backfill TSDB", "user", db.userID, "bucket_dir", db.Dir(), "err", err)
+		}
+	})
+}
+
+func (i *Ingester) compactAllBackfillTSDBs(ctx context.Context) {
+	i.runConcurrentBackfillWorkers(ctx, i.cfg.BlocksStorageConfig.TSDB.ShipConcurrency, func(db *userTSDB) {
+		h := db.Head()
+		if err := db.CompactHead(tsdb.NewRangeHead(h, h.MinTime(), h.MaxTime())); err != nil {
+			level.Error(util.Logger).Log("msg", "unable to compact backfill TSDB", "user", db.userID, "bucket_dir", db.Dir(), "err", err)
+		}
+	})
+}
+
+func (i *Ingester) shipAllBackfillTSDBs(ctx context.Context) {
+	i.runConcurrentBackfillWorkers(ctx, i.cfg.BlocksStorageConfig.TSDB.ShipConcurrency, func(db *userTSDB) {
+		if db.shipper == nil {
+			return
+		}
+		if uploaded, err := db.shipper.Sync(context.Background()); err != nil {
+			level.Warn(util.Logger).Log("msg", "shipper failed to synchronize backfill TSDB blocks with the storage", "user", db.userID, "uploaded", uploaded, "bucket_dir", db.Dir(), "err", err)
+		} else {
+			level.Debug(util.Logger).Log("msg", "shipper successfully synchronized backfill TSDB blocks with storage", "user", db.userID, "uploaded", uploaded, "bucket_dir", db.Dir())
+		}
+	})
+}
+
+func (i *Ingester) runConcurrentBackfillWorkers(ctx context.Context, concurrency int, userFunc func(*userTSDB)) {
```

Comment:
I think `runConcurrentBackfillWorkers()` and `runConcurrentUserWorkers()` could be generalised into a single function like this:

```
func runConcurrentUserWorkers(ctx context.Context, userIDs []string, concurrency int, userFunc func(userID string))
```

Then:
- You need two functions to generate the list of user IDs. We already have `i.getTSDBUsers()` then you could do the same for the backfill TSDBs.
- The callback passed to `runConcurrentBackfillWorkers()` but looks easy passing the userID to the callback and then having the callback fetching the backfilling TSDBs. I understand the actual parallelisation is not the exact same, but we should strive for simplicity first (without compromising correctness) and then we can always optimise it. As already mentioned a bunch of times, I expect backfilling to be an uncommon use case (I don't want to keep multiple TSDBs open for every tenant any time)

_This is a refactoring that could be done in a preliminary PR._